cli/cmd/agent_aws-install_ec2ssm.go

```diff
@@ -0,0 +1,249 @@
+		"tag_key", "", "only install agents on infra with this tag key set",
+	)
+	agentInstallAWSSSMCmd.Flags().StringSliceVar(&agentCmdState.InstallTag,
+		"tag", []string{}, "only install agents on infra with this tag",
+	)
+	agentInstallAWSSSMCmd.Flags().StringSliceVarP(&agentCmdState.InstallIncludeRegions,
+		"include_regions", "r", []string{}, "list of regions to filter on",
+	)
+	agentInstallAWSSSMCmd.Flags().StringVar(&agentCmdState.InstallAgentToken,
+		"token", "", "agent access token",
+	)
+	agentInstallAWSSSMCmd.Flags().IntVarP(
+		&agentCmdState.InstallMaxParallelism,
+		"max_parallelism",
+		"n",
+		50,
+		"maximum number of workers executing AWS API calls, set if rate limits are lower or higher than normal",
+	)
+	agentInstallAWSSSMCmd.Flags().StringVar(
+		&agentCmdState.InstallBYORole,
+		"iam_role_name",
+		"",
+		"IAM role name (not ARN) with SSM policy, if not provided then an ephemeral role will be created",
+	)
+}
+
+func installAWSSSM(_ *cobra.Command, _ []string) error {
+	token := agentCmdState.InstallAgentToken
+	if token == "" {
+		if cli.InteractiveMode() {
+			// user didn't provide an agent token
+			cli.Log.Debugw("agent token not provided, asking user to select one now")
+			var err error
+			token, err = selectAgentAccessToken()
+			if err != nil {
+				return err
+			}
+		} else {
+			return errors.New("user did not provide or interactively select an agent token")
+		}
+	}
+
+	runners, err := awsDescribeInstances()
+	if err != nil {
+		return err
+	}
+
+	cfg, err := GetConfig()
+	if err != nil {
+		return err
+	}
+	role, instanceProfile, err := SetupSSMAccess(cfg, agentCmdState.InstallBYORole, token)
+	defer func() {
+		err := TeardownSSMAccess(cfg, role, instanceProfile, agentCmdState.InstallBYORole) // clean up after ourselves
+		cli.Log.Warnw("got an error while tearing down IAM infra", "error", err)
+	}()
+	if err != nil {
+		return err
+	}
+
+	wg := new(sync.WaitGroup)
+	wp := workerpool.New(agentCmdState.InstallMaxParallelism)
+	for _, runner := range runners {
+		wg.Add(1)
+
+		// In order to use `wp.Submit()`, the input func() must not take any arguments.
+		// Copy the runner info to dedicated variable in the goroutine to prevent race overwrite
+		runnerCopyWg := new(sync.WaitGroup)
+		runnerCopyWg.Add(1)
+
+		wp.Submit(func() {
+			defer wg.Done()
+
+			threadRunner := *runner
+			runnerCopyWg.Done()
+
+			cli.Log.Debugw("runner info: ",
+				"user", threadRunner.Runner.User,
+				"region", threadRunner.Region,
+				"az", threadRunner.AvailabilityZone,
+				"instance_id", threadRunner.InstanceID,
+				"hostname", threadRunner.Runner.Hostname,
+			)
+
+			// Attach an instance profile with our new role to the runner
+			err = threadRunner.AssociateInstanceProfileWithRunner(cfg, instanceProfile)
+			if err != nil {
+				cli.Log.Debugw("failed to attach instance profile to runner",
+					"error", err,
+					"instance ID", threadRunner.InstanceID,
+					"role", role,
+					"instance profile", instanceProfile,
```

Comment:
light nit, TIOLI: logs are usually easier to parse if labels don't have spaces in them.